### ISSUE TITLE
Cahing for a custom FeatureDefinitionProvider

### ISSUE THREAD

**Issue by tahazayed** [microsoft/FeatureManagement-Dotnet#367]
I am building a custom FeatureDefinitionProvider and I want to support caching for it, but this interface IFeatureDefinitionProviderCacheable is internal; would you please change it to be public?

**Comment by tahazayed** [comment]
I am using the same way as tests/Tests.FeatureManagement/InMemoryFeatureDefinitionProvider.cs

**Comment by thompson-tomo** [comment]
You shouldn't need to add that interface to your provider given it is simply to alter handling for test cases and is not needed in actual use case. Please try it without and also what i would be doing is, implementing your provider as a configuration provider that way you definitions are immediately cached for the library to use.

**Comment by jimmyca15** [comment]
@tahazayed Indeed this functionality should be available to you.

@rossgrambo [this conversation](https://github.com/microsoft/FeatureManagement-Dotnet/pull/229#discussion_r1189033113) mentions a plan to open up this functionality and to file an issue. Are we tracking it?

**Comment by rossgrambo** [comment]
No, I don't believe it was ever tracked. I'll treat this as the tracking task and look into it.

**Comment by tahazayed** [comment]
@jimmyca15 , @rossgrambo Thank you. In my circumstances, the alternative SaaS dynamic feature management costs $20,000 per month.

**Comment by rossgrambo** [comment]
Hey @tahazayed, thanks for letting us know. I want to align on our different cacheing to ensure you're getting the caching you want.

IFeatureDefinitionProviderCacheable is an interface that enables a specific cache. This cache is for the binding of IConfiguration to a specific class. For our TimeWindowFilter- this cache prevents [IConfiguration.Get<TimeWindowFilterSettings>()](https://github.com/microsoft/FeatureManagement-Dotnet/blob/main/src/Microsoft.FeatureManagement/FeatureFilters/TimeWindowFilter.cs#L36) from firing too much, as this type conversion is expensive a few ms, and can be reduced to roughly ~1/100th of the time with this cache.

Separately, there's the more common caching concern- Feature Flags from a server being supplied to the Feature Management library. FeatureManagement reads flags though the IFeatureDefinitionProvider injected dependency. FeatureManagement does not handle caching in this way, instead, it will always request feature defintions directly from the provider. The provider is in charge of caching or to pull from a cache. The provided ConfigurationFeatureDefinitionProvider pulls from IConfiguration, which is essentially our cache for the flag data. Our server provider ([Azure App Configuration's .NET Provider](https://github.com/Azure/AppConfiguration-DotnetProvider)), updates the Configuration in order to dynamically update flags. This is what @thompson-tomo was referring to when they said the "config layer would act as the caching". 

Being described like this, @tahazayed, which caching were you hoping to take advantage of? 

I'll work on making IFeatureDefinitionProviderCacheable public / opening up that performance gain caching either way, but I don't want you to wait on it if it's not the cache you care about.

**Comment by tahazayed** [comment]
Thank you for letting me know. I think I want to use what @thompson-tomo is referring to. Do you have an example of how to inject a custom configuration definition provider?

**Comment by thompson-tomo** [comment]
@tahazayed easiest way i have found is to follow the example from Microsoft: https://learn.microsoft.com/en-us/dotnet/core/extensions/custom-configuration-provider and look at some of the samples around.

**Comment by tahazayed** [comment]
I wonder where the cache configuration is and when it became invalid, as it seems after implementation, the cache lasts forever

**Comment by thompson-tomo** [comment]
configuration never expires, it is up to the configuration provider to implement an update/expiration mechanism

**Comment by tahazayed** [comment]
What is value using it? i get the same result as using Appsettings.json

**Comment by thompson-tomo** [comment]
Value of using a custom configuration provider? Benefit is you can source your feature flags from an external system with the config provider doing the transformation to a dictionary hence making it usable.

**Comment by tahazayed** [comment]
I can achieve the same with a custom FeatureDefinitionProvider as I have to implement a cachig with expiration myself anyway,
Also, if I use a configuration provider, I have to hide a property called "Data" from the base class ConfigurationProvider to get the caching work

**Comment by thompson-tomo** [comment]
I don't think it is as complex as you are thinking. Perhaps take a look at: https://learn.microsoft.com/en-us/azure/azure-app-configuration/enable-dynamic-configuration-dotnet-core?tabs=core6x

**Comment by tahazayed** [comment]
I have tried it; I had to write three classes instead of one.

**Comment by jpolvora** [comment]
I have been using this library a lot for over a year in projects that I work on. Over time, several challenges have arisen in the implementation and adaptation to business requirements, and after studying the code extensively, I decided to experiment and play with the idea of ​​creating a custom library. In the end, I ended up creating an extension/complement library, which adds more common features that have generally been applied in the projects that I have worked on. In a way, some of the features implemented answer several of the questions raised here. I thought it would be interesting to share it with the community. If anyone wants to take a look, the project is available here on Github, under the name [Microsoft.FeatureManagement.Plus](https://github.com/jpolvora/Microsoft.FeatureManagement.Plus)

**Comment by rthomas-rdm** [comment]
@jimmyca15 @rossgrambo 
Would you accept a PR to update the IFeatureDefinitionProviderCacheable to be public?

I have a use case where we want to write a very simple wrapper around the out-of-the-box ConfigurationFeatureDefinitionProvider. Without this our simple wrapper is unable to leverage the existing caching mechanism forcing us to write unnecessary extra code if we want to implement the caching by also maintaining a separate copy of the [FeatureManager where the actual parameter caching happens](https://github.com/microsoft/FeatureManagement-Dotnet/blob/b9907a0213c034560957aedf756a64ac95f15c95/src/Microsoft.FeatureManagement/FeatureManager.cs#L683).

**Comment by jimmyca15** [comment]
The use case is clear, extend the existing IFeatureDefinitionProvider. I don't see a problem to make it public.

For my reference, the access level discussion in the initial PR is [here](https://github.com/microsoft/FeatureManagement-Dotnet/pull/229#discussion_r1189033113).

**Comment by jimmyca15** [comment]
@rthomas-rdm 

I had the chance to discuss this with a few folks and have a better picture than I did during my previous comment. Making this interface public may not be the best bet. This original issue here shows that the vagueness of the interface makes it difficult to determine what exactly it does.

Instead, it may be better to expose a switch on `FeatureManager` such as `public bool FilterParameterCachingEnabled { get; set; }`, or in FeatureManagementOptions. For backward compatibility, the default would be false, and our configuration provider would ignore the setting.

There is still a concern of whether this switch is clear enough.

**Comment by RyanThomas73** [comment]
@jimmyca15 

> RE: There is still a concern of whether this switch is clear enough.

I believe we should be able to further mitigate this with an explicit doc comment explaining the behavior in more detail.

> For backward compatibility, the default would be false, and our configuration provider would ignore the setting.

Assuming we went with a property on FeatureManagementOptions - for the backwards compatibility support would the maintainers prefer:

```csharp
            // Option A - Keep the interface for the backwards compatibility behavior check:
            var isFilterParameterCachingEnabled = _options.IsFeatureManagement.FilterParameterCachingEnabled 
                || (_featureDefinitionProvider is IFeatureDefinitionProviderCacheable);

            // Option B - Eliminate the confusing interface and just key off the actual default concrete type.
            var isFilterParameterCachingEnabled = _options.IsFeatureManagement.FilterParameterCachingEnabled 
                || (_featureDefinitionProvider is ConfigurationFeatureDefinitionProvider);   
```

Option B would be less confusing and more self documenting in my opinion. But option B would require a bigger update since it looks like automated test cases would need to be updated to use the new options property in stead of specifying the interface on the InMemoryFeatureDefinitionProvider class.

**Comment by jimmyca15** [comment]
@RyanThomas73 

I've come up with an alternative solution which should address all the concerns that were raised, and nothing will have to be made public. It aims to address the 'wrapper' scenario you have explicitly.

I need to confirm though -- 
1. when you wrap the built-in `ConfigurationFeatureDefinitionProvider`, do you touch/modify any [`FeatureFilterConfiguration.Parameters`](https://github.com/microsoft/FeatureManagement-Dotnet/blob/b9907a0213c034560957aedf756a64ac95f15c95/src/Microsoft.FeatureManagement/FeatureFilterConfiguration.cs#L22-L23) in your wrapper?
2. Do you serve the same `FeatureDefinition` provided by the contained `ConfigurationFeatureDefinitionProvider`?

If you're using the same `Parameters` passed by the built-in provider, we can simply update the cache check for our internal `ConfigurationWrapper` type, and you will get the performance benefits without having to make any code changes on your side.

**Comment by kimsey0** [comment]
@jimmyca15, I'd like caching for a completely custom `IFeatureDefinitionProvider`, not a wrapper around `ConfigurationFeatureDefinitionProvider`. Either making the interface public or the `FilterParameterCachingEnabled` option would work for my use case, but the `ConfigurationWrapper` approach would not.

We currently use `ConfigurationFeatureDefinitionProvider` solely to get filter parameter caching, even though our feature definitions come from blob storage in a format we could map directly to `FeatureDefinition` objects. This forces us to flatten our native format into configuration keys, only for `ConfigurationFeatureDefinitionProvider` to parse them back (plus a `GetChildKeys` precomputation hack to work around the enumeration performance cost of going through `IConfigurationProvider`). With caching available to custom providers, we could just hold a `Dictionary<string, FeatureDefinition>` and skip the entire configuration layer.

**Comment by jimmyca15** [comment]
@kimsey0 

For that scenario, I think the enhancement being added in this PR we are working on at the moment may work for you.

https://github.com/microsoft/FeatureManagement-Dotnet/pull/589

The PR description has an overview, and before & after code snippet. You'll no longer have to go from your database format to IConfiguration.

**Comment by kimsey0** [comment]
Thanks! I hadn't seen that pull request, but it's a much better solution for our use case.

**Comment by RyanThomas73** [comment]
> [@RyanThomas73](https://github.com/RyanThomas73)
> 
> I've come up with an alternative solution which should address all the concerns that were raised, and nothing will have to be made public. It aims to address the 'wrapper' scenario you have explicitly.
> 
> I need to confirm though --
> 
> 1. when you wrap the built-in `ConfigurationFeatureDefinitionProvider`, do you touch/modify any [`FeatureFilterConfiguration.Parameters`](https://github.com/microsoft/FeatureManagement-Dotnet/blob/b9907a0213c034560957aedf756a64ac95f15c95/src/Microsoft.FeatureManagement/FeatureFilterConfiguration.cs#L22-L23) in your wrapper?
> 2. Do you serve the same `FeatureDefinition` provided by the contained `ConfigurationFeatureDefinitionProvider`?
> 
> If you're using the same `Parameters` passed by the built-in provider, we can simply update the cache check for our internal `ConfigurationWrapper` type, and you will get the performance benefits without having to make any code changes on your side.

@jimmyca15 We do not mutate any .Parameters and we do return the the same FeatureDefinition returned by the underlying ConfigurationFeatureDefinitionProvider. You alternative solution would likely meet our needs just fine.

**Comment by RyanThomas73** [comment]
@jimmyca15 

We appreciate the update for the cache check that was added.
Is there a road map page or informal ETA you can provide for when the next planned package version release containing this change will go out?

**Comment by jimmyca15** [comment]
@RyanThomas73 Shooting to get it out this week

**Comment by jimmyca15** [comment]
These improvements have been released in version 4.5.0

[release notes](https://github.com/Azure/AppConfiguration/blob/main/releaseNotes/Microsoft.Featuremanagement.md)
